### PR TITLE
Issue #1602 - Deployment Error should fail fast

### PR DESCRIPTION
**Pull request by joakime**
Fixes deployment errors that occur during WebAppContext.doStart() to respond on Throwable, not Exception.  Allowing for a fast fail of the deployment and not running subsequent configurations and or SCI's if one fails.

### CODE REVIEW

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java [hunk 1]
```diff
@@ -40,7 +40,6 @@
 import org.eclipse.jetty.util.log.Log;
 import org.eclipse.jetty.util.log.Logger;
 import org.eclipse.jetty.webapp.AbstractConfiguration;
-import org.eclipse.jetty.webapp.Configuration;
 import org.eclipse.jetty.webapp.WebAppClassLoader;
```
Comment (janbartel) [inline jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java]: Not a huge fan of making incidental  changes in other files, although as this is just a single line, ok.

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java [hunk 2]
```diff
@@ -548,14 +548,14 @@ protected void doStart() throws Exception
             if (isLogUrlOnStart())
                 dumpUrl();
         }
-        catch (Exception e)
+        catch (Throwable t)
         {
```
Comment (janbartel) [inline jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java]: Definitely fixes the problem when throwUnavailableOnStartupException==false. However, when it's true, the problem is back again!

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java (hunk 1 above)
Comment (joakime) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java]: I was digging into EnvConfiguration to see if it was doing anything strange with exceptions, didn't notice the removal of the unusued import.  I can undo easily enough.

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java (hunk 2 above)
Comment (joakime) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java]: Yet oddly, `setThrowUnavailableOnStartupException(true);` and `setThrowUnavailableOnStartupException(false);` have the same results.  
The deployment fails and is not tried again by the DeploymentManager as the webapp isn't "started" or "available".

Comment (janbartel) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java]: So if the webapp is marked as available==false, and the DM doesn't retry it, why is the ContextHandlerCollection startup trying to start it, only iff the UnavailableException is thrown?

Comment (joakime) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java]: I added to the tests.

There's a scenario for `setThrowUnavailableOnStartupException(true);` and `setThrowUnavailableOnStartupException(false);` now (declared in XML actually)

The results ...

Context | ThrowOnUnavail...Exception | isStarted | isFailed | isAvailable
-------- | ----------------------------------------- | --------- | ------- | ----------
`/badapp` | true | false | true | false
`/badapp-uaf` | false | true | false | false

In both cases, the webapp is unavailable.
The TrackedConfiguration verifies that the Configuration is only attempted once, and fails half-way through.
The ContextHandlerCollection doesn't try to start it again.
The DeploymentManager doesn't try to start it again.

I could (in the testcase) add an actual GET request against the context's to see if they are truly unavailable (it just feels like that kind of test is done elsewhere already)

File: jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java (hunk 1 above)
Comment (joakime) [reply jetty-plus/src/main/java/org/eclipse/jetty/plus/webapp/EnvConfiguration.java]: Actually, not that easy to undo. :-)
The editor refuses to let me add back an import that isn't used, even on revert!

File: jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java (hunk 2 above)
Comment (joakime) [reply jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebAppContext.java]: ContextHandlerCollection startup (doStart()) has no handlers when it executes.

The ContextHandler's are added by the DeploymentManager Bindings to the post-started ContextHandlerCollection.

The Bindings are as follows ...

Binding | Role
-------- | ----
StandardDeployer | ContextHandlerCollection.addHandler(contextHandler)
StandardStarter | ContextHandlerCollection.manage(contextHandler)
StandardStopper | ContextHandlerCollection.unmanage(contextHandler)
StandardUndeployer | ContextHandlerCollection.getHandlers() // remove context recursively

The failures are occurring during StandardStarter.processBinding().

**Comment by joakime** [comment]
@gregw requested change applied, see updated PR #2110

**Comment by joakime** [comment]
@gregw with the change you requested, server startup has changed. (not sure if this is for the better)

If you have setThrowUnavailableOnStartupException(true); the server will now fail startup and stop (prior to this change, this didn't happen.  The context would just not be present.)

If you have setThrowUnavailableOnStartupException(false); the server will continue to run, and the context will return 503.

If a hot deployment introduces a failure, the server will stop/shutdown now.

**Comment by gregw** [comment]
@joakime 
failing to start with setThrowUnavailableOnStartupException(true) is indeed different behaviour, but I consider it fixing a bug, so I'm good with that.

However, the server should not exit on hot deploy?  What is the mechanism for that? My understanding is that on hot deploy the handler will be added and start will be called from within addBean. If an exception is thrown, it is caught and potentially wrapped as a RuntimeException and then rethrown.  But it will then only go back to the DeploymentManager component that called the addHandler, which itself should be protected by a try/catch, and it should just log.warn the exception and carry on (perhaps marking the context as unmanaged?).     It should not stop the server.

**Comment by joakime** [comment]
@gregw I need to update the testcase anyway, once the intended behavior is well described.

So far, I can see the following tests ...

1. Bad SCI in webapp present on server start.
  **a.** setThrowUnavailableOnStartupException(true); - server doesn't start
  **b.** setThrowUnavailableOnStartupException(false); - server starts, webapp is present in ContextHandlerCollection, is started && !failed && !available.  requests to webapp result in 503 response.

2. Bad SCI introduced via Hot Deploy.
  **a.** setThrowUnavailableOnStartupException(true); - server starts normally, hot deploy fails, server is still running, webapp is not (?) present in ContextHandlerCollection?
  **b.** setThrowUnavailableOnStartupException(true); - server starts normally, hot deploy fails, server is still running, webapp is present in ContextHandlerCollection, is !started && failed && !available, requests to webapp result in 404 response.
  **c.** setThrowUnavailableOnStartupException(false); - server starts, hot deploy fails, server is still running, webapp is present in ContextHandlerCollection, is started && !failed && !available.  requests to webapp result in 503 response.

of use case 2.a and 2.b, which is correct?

**Comment by gregw** [comment]
@joakime,

It's not just bad SCI. It's anytime a context throws.  This used to be if a RuntimeException was thrown OR and Error was thrown OR setThrowUnavailableOnStartupException(true).

So the first fix is a good one... ie the context now catches Throwable so the dual start issue happens on if setThrowUnavailableOnStartupException(true).

Your 1.a and 1.b behaviours are good.

I think we have some options on 2.x

2.a setThrowUnavailableOnStartupException(true); - server starts normally, hot deploy catches exception and context is now in FAILED state.   The deployer now has a choice of what to do... it could remove it from the ContextHandlerCollection, but that might make debug difficult... or it can just leave it added and failed (note that available status doesn't matter as failed contexts will not handle any requests so availability is not check).    I think leaving it added and failed is best.    Potentially it could also be unmanaged, so it would just remain as a placeholder of what happened.

2.b setThrowUnavailableOnStartupException(false); - server starts, hot deploy DOES NOT FAIL as the exception is caught, server is still running, webapp is present in ContextHandlerCollection, is started && !failed && !available. requests to webapp result in 503 response.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java [hunk 3]
```diff
@@ -35,12 +36,17 @@
...
+
         ContextHandler handler = app.getContextHandler();
-        
-        // start the handler
-        handler.start();
-        
-        // After starting let the context manage state
-        app.getDeploymentManager().getContexts().manage(handler);
+
+        if (contexts.isRunning() && !handler.isRunning())
```
Comment (gregw) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: Change to 
```
if (contexts.isStarted() && handler.isStopped())
```
so failed handlers are not started.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java [hunk 4]
```diff
@@ -35,12 +36,17 @@
...
+        {
+            // start the handler manually
```
Comment (gregw) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: Do a `LOG.warn` here that StandardStarter may be removed in a future release.

**Comment by janbartel** [comment]
Where are we on this one?

**Comment by gregw** [comment]
My strong preference is that we update the StandardStarter to do the:
```
    if (contexts.isStarted() && handler.isStopped())
```
test.   We then review StandardStarter for jetty-10.

**Comment by gregw** [comment]
@joakime can you update this PR as recommended.  I really think that is the best short term solution as it keeps the StandardStarter.  We can review the need for it in jetty-10.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java (hunk 3 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: Current test cases for this PR show that failed handlers are not restarted again.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java (hunk 4 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: out of scope for this PR.
better as a PR for #2116

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java (hunk 3 above)
Comment (gregw) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: However, I believe that the `contexts.isStarted() && handler.isStopped()` form is clearer and more correct. It avoids untested corner cases of starting and stopping states.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java (hunk 4 above)
Comment (gregw) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: far enough, so let's just change the test and get this committed.

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java (hunk 3 above)
Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: `context.isRunning()` seems more appropriate as its for both STARTED and STARTING.

Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: Tests with `handler.isStopped()` are causing a deployment loop if `WebAppContext.isThrowUnavailableOnStartupException == false` now.
Troubleshooting ...

Comment (joakime) [reply jetty-deploy/src/main/java/org/eclipse/jetty/deploy/bindings/StandardStarter.java]: Seems that the handler with `isThrowUnavailableOnStartupException == false` goes in state ...

1. STARTING
2. (error occurs, but is not reported due to `isThrowUnavailableOnStartupException == false`)
3. STOPPED

If the same handler with setting `isThrowUnavailableOnStartupException == true` then it seems to go into ...

1. STARTING
2. (error occurs, flows out and is caught by state handling)
3. FAILED

**Comment by joakime** [comment]
@gregw lots of updates in 48123ee.  your requested changes, and a bunch more around testing a variety of deployment scenarios

File: jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java [hunk 5]
```diff
@@ -507,6 +507,17 @@ private void requestAppGoal(AppEntry appentry, String nodeName)
...
+            // migrate to FAILED node
+            Node failed = _lifecycle.getNodeByName(AppLifeCycle.FAILED);
+            appentry.setLifeCycleNode(failed);
+            try
+            {
+                _lifecycle.runBindings(failed, appentry.app, this);
+            }
+            catch (Throwable ignore)
+            {
+                // The runBindings failed for 'failed' node, no point doing anything else here.
```
Comment (gregw) [inline jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentManager.java]: I think there should be at least a LOG.ignore(ignore) here.

**Pull request merged**